academy/ibc-dev/ibcmw-integrate.md

```diff
@@ -0,0 +1,156 @@
+
+    ```go
+    //app.go
+
+    // create a keeper for the stateful middleware
+    mw1Keeper := mw1.NewKeeper(storeKey1)
+    mw3Keeper := mw3.NewKeeper(storeKey3)
+
+    // instantiate the middleware as IBCModules
+    mw1IBCModule := mw1.NewIBCModule(mw1Keeper)
+    mw2IBCModule := mw2.NewIBCModule() // middleware2 is stateless middleware
+    mw3IBCModule := mw3.NewIBCModule(mw3Keeper)
+
+    // register middleware in app module
+    // if the module maintains independent state and/or processes sdk.Msgs
+    app.moduleManager = module.NewManager(
+        ...
+        mw1.NewAppModule(mw1Keeper),
+        mw3.NewAppModule(mw3Keeper),
+        transfer.NewAppModule(transferKeeper),
+        custom.NewAppModule(customKeeper)
+    )
+    
+    ```
+
+    <ExpansionPanel title="Example code">
+    <!-- TODO: put code above in the expansion panel -->
+    </ExpansionPanel>
+
+- As mentioned in the previous section, in the case where the middleware wishes to send a packet or acknowledgment without the involvement of the underlying application, it should be given access to the same `scopedKeeper` as the base application so that it can retrieve the capabilities by itself.
+
+    ```go
+        //app.go
+
+        // add scoped keepers in case the middleware wishes to 
+        // send a packet or acknowledgment without 
+        // the involvement of the underlying application
+        scopedKeeperTransfer := capabilityKeeper.NewScopedKeeper("transfer")
+        scopedKeeperCustom1 := capabilityKeeper.NewScopedKeeper("custom1")
+        scopedKeeperCustom2 := capabilityKeeper.NewScopedKeeper("custom2")
+        
+    ```
+
+    <ExpansionPanel title="Example code">
+    <!-- TODO: put code above in the expansion panel -->
+    </ExpansionPanel>
+
+- Each application stack must reserve its own unique port with core IBC. Thus two stacks with the same base application must bind to separate ports.
+
+    ```go
+        // initialize base IBC applications
+        //
+        // if you want to create two different stacks with the same base application,
+        // they must be given different scopedKeepers and assigned different ports
+        transferIBCModule := transfer.NewIBCModule(transferKeeper)
+        customIBCModule1 := custom.NewIBCModule(customKeeper, "portCustom1")
+        customIBCModule2 := custom.NewIBCModule(customKeeper, "portCustom2")
+        
+    ```
+
+    <ExpansionPanel title="Example code">
+    <!-- TODO: put code above in the expansion panel -->
+    </ExpansionPanel>
+
+
+- The order of middleware **matters**, function calls from IBC to the application travel from top-level middleware to the bottom middleware and then to the application. Function calls from the application to IBC goes through the bottom middleware in order to the top middleware and then to core IBC handlers. Thus the same set of middleware put in different orders may produce different effects.
```

Comment:
We should maybe specify which calls run IBC -> app (the functions defined in `porttypes.AppModule`), and function calls that run app -> IBC (the functions defined in `ICS4Wrapper`)

Reply:
Ahh I see that you have it defined elsewhere, but maybe its more intuitive if we mention that you can visualize it just by the two interfaces you have to implement?

Reply:
Oh I never considered that the interface splits up as the direction of information flows. Can add that for sure ;)